crates/core/transaction/src/plan/build.rs

```diff
@@ -488,13 +235,149 @@ impl UnauthTransaction {
         }
 
         // Compute the binding signature and assemble the transaction.
-        let binding_signing_key = rdsa::SigningKey::from(self.synthetic_blinding_factor);
-        let auth_hash = self.inner.transaction_body.auth_hash();
+        let binding_signing_key = rdsa::SigningKey::from(synthetic_blinding_factor);
+        let auth_hash = transaction.transaction_body.auth_hash();
+
         let binding_sig = binding_signing_key.sign(rng, auth_hash.as_bytes());
         tracing::debug!(bvk = ?rdsa::VerificationKey::from(&binding_signing_key), ?auth_hash);
 
-        self.inner.binding_sig = binding_sig;
+        transaction.binding_sig = binding_sig;
+
+        Ok(transaction)
+    }
+
+    /// Build the serial transaction this plan describes.
+    /// Arguments
+    ///     - `full_viewing_key`, the [`FullViewingKey`] for the source funds;
+    ///     - `witness_data`, the [`WitnessData`] used for proving;
+    ///     -  `auth_data`, the [`AuthorizationData`] required to sign transaction.
+    pub fn build(
+        self,
+        full_viewing_key: &FullViewingKey,
+        witness_data: WitnessData,
+        auth_data: &AuthorizationData,
+    ) -> Result<Transaction> {
+        // Add the memo.
+        let mut _memo: Option<MemoCiphertext> = None;
+        let mut memo_key: Option<PayloadKey> = None;
+        if self.memo_plan.is_some() {
+            let memo_plan = self
+                .memo_plan
+                .clone()
+                .ok_or_else(|| anyhow!("missing memo_plan in TransactionPlan"))?;
+            _memo = memo_plan.memo().ok();
+            memo_key = Some(memo_plan.key);
+        }
+
+        let mut actions: Vec<Action> = Vec::new();
+
+        // We build the actions sorted by type, with all spends first, then all
+        // outputs, etc.  This order has to align with the ordering in
+        // TransactionPlan::effect_hash, which computes the auth hash of the
+        // transaction we'll build here without actually building it.
```

Comment:
We should just iterate over the list of actions in the plan in the order they're in the plan. I believed that this had already been done as part of earlier `EffectHash` changes, but that seems to not have been the case: https://github.com/penumbra-zone/penumbra/blob/main/crates/core/transaction/src/effect_hash.rs#L142-L146

We need to not compute the `EffectHash` that way for code simplicity and to allow constrained devices like hardware wallets to stream the `ActionPlan`s rather than process a full `TransactionPlan` all at once.

Reply:
I fixed this in c3ec960bd (cc @redshiftzero @cronokirby)

Reply:
removed comment